src/GraphQL.MemoryCache/AutomaticPersistedQueries/AutomaticPersistedQueriesExecution.cs

```diff
@@ -0,0 +1,120 @@
+/// https://www.apollographql.com/docs/react/api/link/persisted-queries/
+/// </summary>
+public class AutomaticPersistedQueriesExecution : IConfigureExecution
+{
+    private readonly IQueryCache _cache;
+    private const string SUPPORTED_VERSION = "1";
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    public AutomaticPersistedQueriesExecution(IQueryCache cache)
+    {
+        _cache = cache;
+    }
+
+    /// <summary>
+    /// Searching APQ fields in <see cref="ExecutionOptions.Extensions"/> based on a protocol:
+    /// https://www.apollographql.com/docs/react/api/link/persisted-queries/#protocol
+    /// </summary>
+    protected virtual (string? Hash, string? Version, bool Enabled) GetAPQProperties(Inputs? extensions)
+    {
+        string? hashResult = null;
+        string? versionResult = null;
+        bool enabled = false;
+
+        if (extensions?.TryGetValue("persistedQuery", out var persistedQueryObject) ?? false)
+        {
+            enabled = true;
+
+            if (persistedQueryObject is Dictionary<string, object> persistedQuery)
+            {
+                if (persistedQuery.TryGetValue("sha256Hash", out var sha256HashObject) && sha256HashObject is string sha256Hash && !string.IsNullOrWhiteSpace(sha256Hash))
+                {
+                    hashResult = sha256Hash;
+                }
+
+                if (persistedQuery.TryGetValue("version", out var versionObject) && versionObject is string version && !string.IsNullOrWhiteSpace(version))
+                {
+                    versionResult = version;
+                }
+            }
+        }
+
+        return (hashResult, versionResult, enabled);
+    }
+
+    /// <summary>
+    /// Compute SHA256 hash.
+    /// </summary>
+    protected virtual string ComputeSHA256(string input)
+    {
+        var bytes = SHA256.Create().ComputeHash(Encoding.UTF8.GetBytes(input));
+
+        var builder = new StringBuilder();
+        foreach (var item in bytes)
+        {
+            builder.Append(item.ToString("x2"));
+        }
+
+        return builder.ToString();
+    }
+
+    /// <summary>
+    /// Create <see cref="ExecutionResult"/> with provided error.
+    /// </summary>
+    protected virtual ExecutionResult CreateExecutionResult(ExecutionError error) => new ExecutionResult { Errors = new ExecutionErrors { error } };
+
+    /// <inheritdoc/>
+    public async Task<ExecutionResult> ExecuteAsync(ExecutionOptions options, ExecutionDelegate next)
+    {
+        var apq = GetAPQProperties(options.Extensions);
+
+        if (apq.Enabled && apq.Version != SUPPORTED_VERSION)
+        {
+            return CreateExecutionResult(new PersistedQueryUnsupportedVersionError(apq.Version));
+        }
+
+        if (string.IsNullOrWhiteSpace(options.Query))
+        {
+            if (string.IsNullOrWhiteSpace(apq.Hash))
+            {
+                return CreateExecutionResult(new ExecutionError("GraphQL query is missing."));
```

Comment:
Let's create a class for this situation and move the check into the `DocumentExecuter.CoreExecuteAsync` method.  This is what I wrote for the server side which probably should be added to the main project.  The other errors you've added probably can inherit from `RequestError` also.  @sungam3r should also review.

```cs
/// <summary>
/// Represents an error indicating that no GraphQL query was provided to the request.
/// </summary>
public class QueryMissingError : RequestError
{
    /// <inheritdoc cref="QueryMissingError"/>
    public QueryMissingError() : base("GraphQL query is missing.") { }
}

/// <summary>
/// Represents an error that occurred prior to the execution of the request.
/// </summary>
public class RequestError : ExecutionError
{
    /// <summary>
    /// Initializes an instance with the specified message.
    /// </summary>
    public RequestError(string message) : base(message)
    {
        Code = GetErrorCode();
    }

    /// <summary>
    /// Initializes an instance with the specified message and inner exception.
    /// </summary>
    public RequestError(string message, Exception? innerException) : base(message, innerException)
    {
        Code = GetErrorCode();
    }

    internal string GetErrorCode()
    {
        var code = ErrorInfoProvider.GetErrorCode(GetType());
        if (code != "REQUEST_ERROR" && code.EndsWith("_ERROR", StringComparison.Ordinal))
            code = code.Substring(0, code.Length - 6);
        return code;
    }
}
```

Reply:
Do we need to do something else in the server after this PR?

Reply:
Yes; we need to allow requests with a blank Query.  Already done in https://github.com/Shane32/GraphQL.AspNetCore3/pull/28 .

Reply:
If we code DocumentExecuter to return a response with an ExecutionError when query is null, then we can eliminate the query-null checking from the server entirely.  But as of now it throws an `InvalidOperationException`.

Reply:
Now I see I think. Do you want to check for `RequestError` type in server and return `HttpStatusCode.BadRequest` in that case?

Reply:
> Represents an error that occurred prior to the execution of the request.

Validation also occures prior to the execution but this error is not about GraphQL validation, it is about client supplied argument on the transport/infrastructure layer.

Reply:
> Now I see I think. Do you want to check for `RequestError` type in server and return `HttpStatusCode.BadRequest` in that case?

That already occurs, generally speaking.  Meaning that any validation error (where `result.Executed == false`) will return a 400 bad request error.  In GraphQL.AspNetCore3 it is configurable, but the default behavior would be to return a 400 error.

> Validation also occures prior to the execution but this error is not about GraphQL validation, it is about client supplied arguments on the transport/infrastructure layer.

Right, so that's why I made a new class `RequestError` - more for 'good housekeeping' rather than any real implementation difference.

It certainly is possible that the server could return 400 for a `RequestError` exception -- even if it returns 200 for validation errors -- if we use this type of inheritance.

Reply:
Personally I think of a request without a query as a validation/syntax error.  Meaning, what's the difference between these?

```json
{
  "query": "",
  "variables": { "stuff": 123 }
}
```

and

```json
{
  "query": null,
  "variables": { "stuff": 123 }
}
```

and

```json
{
  "query": "{",
  "variables": { "stuff": 123 }
}
```

All are essentially the same issue -- the query is invalid.  It's not a problem deserializing the GraphQLRequest json.  It's a problem with the query.  I don't have an issue differentiating those errors, just saying, all are somewhat similar errors.

Reply:
The difference is whether or not we are inside GraphQL engine or outside.

> That already occurs, generally speaking. Meaning that any validation error (where result.Executed == false) will return a 400 bad request error. 

Where is that code?

Reply:
Here is the pointer to the code within my PR https://github.com/graphql-dotnet/server/pull/774 :

https://github.com/graphql-dotnet/server/blob/f2a7011c74ac5317f22fe4149c969d4ee3f241f7/src/Transports.AspNetCore/GraphQLHttpMiddleware.cs#L402

Reply:
Ah, new code. I no longer have the strength to review anything today, only tomorrow.

Reply:
Looks like error classes from #774 are needed for both PRs.